### PR TITLE
[SPARK-16186][SQL] Support partition batch pruning with `IN` predicate in InMemoryTableScanExec

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

One of the most frequent usage patterns for Spark SQL is using **cached tables**. This PR improves `InMemoryTableScanExec` to handle `IN` predicate efficiently by pruning partition batches. Of course, the performance improvement varies over the queries and the datasets. But, for the following simple query, the query duration in Spark UI goes from 9 seconds to 50~90ms. It's about over 100 times faster.

**Before**

``` scala
$ bin/spark-shell --driver-memory 6G
scala> val df = spark.range(2000000000)
scala> df.createOrReplaceTempView("t")
scala> spark.catalog.cacheTable("t")
scala> sql("select id from t where id = 1").collect()    // About 2 mins
scala> sql("select id from t where id = 1").collect()    // less than 90ms
scala> sql("select id from t where id in (1,2,3)").collect()  // 9 seconds
```

**After**

``` scala
scala> sql("select id from t where id in (1,2,3)").collect() // less than 90ms
```

This PR has impacts over 35 queries of TPC-DS if the tables are cached.
Note that this optimization is applied for `IN`.  To apply `IN` predicate having more than 10 items, `spark.sql.optimizer.inSetConversionThreshold` option should be increased.
## How was this patch tested?

Pass the Jenkins tests (including new testcases).

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
cc @rxin , @davies , @cloud-fan .

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala [hunk 1]
```diff
@@ -79,6 +79,11 @@ private[sql] case class InMemoryTableScanExec(
 
     case IsNull(a: Attribute) => statsFor(a).nullCount > 0
     case IsNotNull(a: Attribute) => statsFor(a).count - statsFor(a).nullCount > 0
+
+    case In(a: AttributeReference, list: Seq[Expression])
+      if list.length <= inMemoryPartitionPruningMaxInSize =>
+      list.map(l => statsFor(a).lowerBound <= l.asInstanceOf[Literal] &&
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: where do we make sure the `l` is always literal?

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: Oh, right. I miss that. I'll fix that by checking.
Thank you for review, @cloud-fan !

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: how about we do this optimization for `InSet`? It guarantees the list are all literals and the max length by default is 10. Then we can save the new config.

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: But, that configuration is minimum threshold for InSet. So, the meaning is quite different.

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan .
I updated the PR. IMO,
- InSet is used for large size of `IN` .
- This PR is used for small size of `IN`.

**Comment by dongjoon-hyun** [comment]
Although I decided to make this PR after observing TPC-DS queries, I will definitely update this PR if there are another useful scenarios.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: oh sorry I read the code wrong, yea the config is different.

**Comment by cloud-fan** [comment]
Seems like a reasonable optimization to me, cc @liancheng

**Comment by dongjoon-hyun** [comment]
Thank you, @cloud-fan !

**Comment by davies** [comment]
@dongjoon-hyun Thanks for the patch, this optimization sounds reasonable.

I'm thinking of is it possible to make the optimization for IN/INSET more general. We could have a optimizer to insert a GreaterThanOrEqual and LessThanOrEqual for IN/INSET (checking the data type to make sure they are orderable), it will be cheaper to evaluate. For IN with single expression, we could rewrite as EqualTo. When doing this, we should be careful for null, to respect the sematics of null in IN/INSET. By doing this, all the data sources could benefit from it.

**Comment by davies** [comment]
BTW, we could use constraints to implement this.

**Comment by dongjoon-hyun** [comment]
Thank you for your review and valuable improvement ideas, @davies . Let me rephrase about your ideas,

1. For `IN` with single expression, we definitely had better improve the existing `OptimizeIn` optimizer. (like as you mentioned.)

2. For **sparse** values and small IN, we also can replace `a IN (1,2,3)` into `a = 1 or a = 2 or a = 3`. In terms of generated java code size, the optimized version is smaller.
 ```scala
scala> sql("explain codegen select * from (select explode(array('1','2','3')) a) where a in ('1','2','3')").collect().foreach(println)   // 80 lines
scala> sql("explain codegen select * from (select explode(array('1','2','3')) a) where a = '1' or a = '2' or a ='3'").collect().foreach(println) // 65 lines
```
3. For **consecutive and discretized** values, e.g. 2001, 2002, 2003, ..., 2004, we can improve more cheaper by replacing GreaterThanOrEqual/LessThanOrEqual in Logical Optimizer layer, maybe also `OptimizeIn`?

4. For **ordered partitioned data sources** like InMemory/Parquet, 1~3 will effectively improve the performance.

5. For **the other data sources (unordered, partitioned or not)** and large IN, we need to use INSET as a fallback operation to prevent any regressions.

Are these all what you advise? If I missed something, please comment me.

**Comment by davies** [comment]
For any IN that have more than one expression, we could add another GreaterThanOrEqual/LessThanOrEqual (not replace the IN).

For 2, it's not that obvious yet, we can do that later.

**Comment by dongjoon-hyun** [comment]
Oh, you meant adding additional constraints by using **min** and **max**. I see.

By the way, I have one question. If there are many predicates, does Spark use the predicate in a sorted order?

I'm not sure the newly inserted `GreaterThanOrEqual/LessThanOrEqual` can be used first.

**Comment by davies** [comment]
@dongjoon-hyun There is only single predicate in Filter, it could be AND or OR, so it means we could control the order. For this case, I'm not sure the inserted GreaterThanOrEqual/LessThanOrEqual will come before IN.

**Comment by dongjoon-hyun** [comment]
Okay. Let's summarize before updating PR.
1. In general, `a IN (expression)` will be `a = expression`. `OptimizerIn` optimizer will take care of this.
2. In general, `a IN (2001, 2002, ..., 2009)` will be `2001 <= a AND a <= 2003 AND a IN (2001, 2002, 2009)`. `OptimizerIn` optimizer will take care of this.
Am I correct?

**Comment by davies** [comment]
@dongjoon-hyun Yes, 2) should check the constraints to make it idempotent

**Comment by dongjoon-hyun** [comment]
Hmm. The general idea is good. But, I still think this PR and the idea seem to be complementary to each other.

Sorry, but, if possible, can we proceed that general idea in another PR? 

It's a little bit beyond the title of this PR and the touched files will be different.

**Comment by dongjoon-hyun** [comment]
I'm not sure, but it's just my hope. :)

**Comment by dongjoon-hyun** [comment]
The some of frequent TPC-DS usages were STATE, ZIP, Color strings. The min/max of these values doesn't have much meaning.

**Comment by davies** [comment]
@dongjoon-hyun That's a good point, the current patch is better for performance actually

**Comment by davies** [comment]
Let's go with current patch, I will review it now. Those things could be considered later.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala [hunk 2]
```diff
@@ -79,6 +79,11 @@ private[sql] case class InMemoryTableScanExec(
 
     case IsNull(a: Attribute) => statsFor(a).nullCount > 0
     case IsNotNull(a: Attribute) => statsFor(a).count - statsFor(a).nullCount > 0
+
+    case In(a: AttributeReference, list: Seq[Expression])
+      if list.length <= inMemoryPartitionPruningMaxInSize && list.forall(_.isInstanceOf[Literal]) =>
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: Can we not have this config? Another optimize rule will garantee that the number of expression will not be big.

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: Sure! I'll remove the option related stuff.

**Comment by dongjoon-hyun** [comment]
Hi, @davies .
I removes the option-related stuff from the code/PR description/JIRA description according to your advice.
Thank you for review!

**Comment by davies** [comment]
LGTM, merging this into master.

**Comment by davies** [comment]
Sorry, the jenkins has not finished ...

**Comment by davies** [comment]
Reverted, will merge this again once it passed jenkins.

**Comment by dongjoon-hyun** [comment]
Oh, right. It's pending Jenkins.

**Comment by dongjoon-hyun** [comment]
Maybe, you are confused with https://github.com/apache/spark/pull/13900 .
It passed.

**Comment by dongjoon-hyun** [comment]
Anyway, thank you in advance, @davies . :)

**Comment by dongjoon-hyun** [comment]
Hi, @davies .
Now, it passed.
If there is anything for me to do, please let me know.

**Comment by davies** [comment]
Merged into master, thanks
